WEB-INF/classes/gov/noaa/pfel/erddap/Erddap.java

```diff
@@ -18097,6 +18100,301 @@ public static void makeErddapContentZip(String removeDir, String destinationDir)
+                        }
+                    } catch (Throwable actionT) {
+                        String2.log(subject + "\n" +
+                                "action=" + tAction + "\ncaught:\n" +
+                                MustBe.throwableToString(actionT));
+                    }
+                }
+
+                //trigger RSS action
+                // (after new dataset is in place and if there is either a current or older dataset)
+                if (cooDataset != null && erddap != null)
+                    cooDataset.updateRSS(erddap, change);
+
+            } catch (Throwable subT) {
+                String content = MustBe.throwableToString(subT);
+                String2.log(subject + ":\n" + content);
+                EDStatic.email(EDStatic.emailEverythingToCsv, subject, content);
+            }
+        }
+    }
+
+    /**
+     * This high level method is the entry point to add/remove the
+     * dataset's metadata to/from the proper places in catInfo.
+     *
+     * <p>Since catInfo is a ConcurrentHashMap, this is thread-safe to the extent
+     * that data structures won't be corrupted;
+     * however, it is still susceptible to incorrect information if 2+ thredds
+     * work with the same datasetID at the same time (if one adding and one removing)
+     * because of race conditions.
+     *
+     * @param add determines whether datasetID references will be ADDed or REMOVEd
+     * @param catInfo the new categoryInfo hashMap of hashMaps of hashSets
+     * @param edd the dataset who's info should be added to catInfo
+     */
+    protected static void addRemoveDatasetInfo(boolean add,
+                                            ConcurrentHashMap catInfo, EDD edd) {
+
+        //go through the gridDatasets
+        String id = edd.datasetID();
+
+        //globalAtts
+        categorizeGlobalAtts(add, catInfo, edd, id);
+
+        //go through data variables
+        int nd = edd.dataVariables().length;
+        for (int dv = 0; dv < nd; dv++)
+            categorizeVariableAtts(add, catInfo, edd.dataVariables()[dv], id);
+
+        if (edd instanceof EDDGrid eddGrid) {
+            //go through axis variables
+            int na = eddGrid.axisVariables().length;
+            for (int av = 0; av < na; av++)
+                categorizeVariableAtts(add, catInfo, eddGrid.axisVariables()[av], id);
+        }
+    }
+
+    /**
+     * If useLuceneSearchEngine, this will update the Lucene indices for these datasets.
+     *
+     * <p>Since luceneIndexWriter is thread-safe, this is thread-safe to the extent
+     * that data structures won't be corrupted;
+     * however, it is still susceptible to incorrect information if 2+ thredds
+     * work with the same datasetID at the same time (if one adding and one removing)
+     * because of race conditions.
+     *
+     * @param datasetIDs
+     */
+    protected static void updateLucene(Erddap erddap, StringArray datasetIDs) {
```

Comment:
What about the helper function `addRemoveDatasetInfo` can we keep that as static or not?

Reply:
That should also probably be not-static. I believe everywhere it's called, the parameter `ConcurrentHashMap catInfo` is set to erddap.categoryInfo and so that parameter is likely unnescary if `addRemoveDatasetInfo` is a member of Erddap.